src/signature.rs

```diff
@@ -0,0 +1,167 @@
+use std::borrow::Borrow;
+use std::fmt::Formatter;
+use std::marker::PhantomData;
+use std::ops::{Add, AddAssign, Div, Mul, MulAssign, Neg, Shl, Sub, SubAssign};
+use std::str::FromStr;
+
+use num::bigint::ParseBigIntError;
+use num::BigUint;
+use num::pow;
+
+use crate::{Expression, GadgetBuilder, BooleanExpression, EdwardsCurve, Curve, CurvePoint, EdwardsPointExpression, CompressionFunction};
+use crate::field::{Element, Field};
+
+pub trait SignatureExpression<F: Field, C: EdwardsCurve<F>, CF> {
```

Comment:
Will this require all signatures to use TEC or should we use another name or trait meta?

Reply:
I think that we probably do want to support non-TEC (and non-ECC) signatures, so changing the trait/name is probably a good idea. Maybe EdwardsSignatureExpression? I was going to wait until we'd implemented another signature scheme to refactor, but happy to do it now as well.

Reply:
I think it would be ideal if we could generalize the Schnorr implementation to work with any embedded group. Our current APIs wouldn't support that, but we could add a trait which just performs a group operation in r1cs.